src/build.h

```diff
@@ -147,15 +151,19 @@ struct CommandRunner {
     bool success() const { return status == ExitSuccess; }
   };
   /// Wait for a command to complete, or return false if interrupted.
-  virtual bool WaitForCommand(Result* result) = 0;
+  /// If more_ready is true then the optional TokenPool is monitored too
+  /// and we return when a token becomes available.
+  virtual bool WaitForCommand(Result* result, bool more_ready) = 0;
```

Comment:
The naming of  that method is no longer exactly accurate with that extension.

But not just that - something feels fishy about merging these functionalities into a single method. It's two distinct states mangled into one - *either* you are waiting for jobs to complete as you have an empty work-queue *or* you are waiting for for jobs to complete as you need fresh tokens.

It just happens that those two states share a single wait condition when there is no external job server, but they are simply not the same when there is.

So wrapping those two states into distinct, named wait methods would improve readability.

Reply:
Looks like not making that distinction led you to fusing both sets of FDs into one list in `SubprocessSet`.

You didn't need to merge those sets, and thus could just have kept this hidden within `TokenPool`.

Reply:
... not sure if what I assumed is correct. As it's this process which has to return a token in the case of blocking itself, it does need to await both sets at once after..